components/script/dom/compressionstream.rs

```diff
@@ -0,0 +1,277 @@
+
+        // Step 2. Set this’s format to format.
+        // Step 5. Set this’s transform to a new TransformStream.
+        let transform = TransformStream::new_with_proto(global, None, can_gc);
+        let this = CompressionStream::new_with_proto(global, proto, &transform, format, can_gc);
+
+        // Step 3. Let transformAlgorithm be an algorithm which takes a chunk argument and runs the
+        // compress and enqueue a chunk algorithm with this and chunk.
+        // Step 4. Let flushAlgorithm be an algorithm which takes no argument and runs the compress
+        // flush and enqueue algorithm with this.
+        let transformer_type = TransformerType::Compressor(this.clone());
+
+        // Step 6. Set up this’s transform with transformAlgorithm set to transformAlgorithm and
+        // flushAlgorithm set to flushAlgorithm.
+        let cx = GlobalScope::get_cx();
+        transform.set_up(cx, global, transformer_type, can_gc)?;
+
+        Ok(this)
+    }
+
+    /// <https://streams.spec.whatwg.org/#dom-generictransformstream-readable>
+    fn Readable(&self) -> DomRoot<ReadableStream> {
+        // The readable getter steps are to return this’s transform.[[readable]].
+        self.transform.get_readable()
+    }
+
+    /// <https://streams.spec.whatwg.org/#dom-generictransformstream-writable>
+    fn Writable(&self) -> DomRoot<WritableStream> {
+        // The writable getter steps are to return this’s transform.[[writable]].
+        self.transform.get_writable()
+    }
+}
+
+/// <https://compression.spec.whatwg.org/#compress-and-enqueue-a-chunk>
+pub(crate) fn compress_and_enqueue_a_chunk(
+    cx: SafeJSContext,
+    global: &GlobalScope,
+    cs: &CompressionStream,
+    chunk: SafeHandleValue,
+    controller: &TransformStreamDefaultController,
+    can_gc: CanGc,
+) -> Fallible<()> {
+    // Step 1. If chunk is not a BufferSource type, then throw a TypeError.
+    let conversion_result =
+        ArrayBufferViewOrArrayBuffer::safe_from_jsval(cx, chunk, ()).map_err(|_| {
+            Error::Type("Unable to convert chunk into ArrayBuffer or ArrayBufferView".to_string())
+        })?;
+    let buffer_source = conversion_result.get_success_value().ok_or_else(|| {
+        Error::Type("Unable to convert chunk into ArrayBuffer or ArrayBufferView".to_string())
+    })?;
+    let chunk = match buffer_source {
+        ArrayBufferViewOrArrayBuffer::ArrayBufferView(view) => view.to_vec(),
+        ArrayBufferViewOrArrayBuffer::ArrayBuffer(buffer) => buffer.to_vec(),
+    };
+
+    // Step 2. Let buffer be the result of compressing chunk with cs’s format and context.
+    // NOTE: In our implementation, the enum type of context already indicates the format.
+    let mut compressor = cs.context.borrow_mut();
+    let offset = compressor.get_ref().len();
+    compressor
+        .write_all(&chunk)
+        .map_err(|_| Error::Type("Fail to write all to compressor".to_string()))?;
```

Comment:
```suggestion
        ..map_err(|_| Error::Type("CompressionStream: write_all() failed".to_string()))?;
```

Reply:
Apply similarly for flush, try_finish, and the decompressor paths, this is  helpful